politeiawww/user.go

```diff
@@ -2333,3 +2349,96 @@ func (p *politeiawww) initPaywallChecker() error {
 	go p.checkForPayments()
 	return nil
 }
+
+func (p *politeiawww) totpCheck(code string, u *user.User) error {
```

Comment:
I still find this function is hard to follow due to the nesting.  I took a quick crack at trying to flatten it out.  Would something like this be easier to read?

```
func (p *politeiawww) totpCheck(code string, u *user.User) error {
	// Verify a code was provided
	if code == "" {
		log.Debugf("login: totp code required %v", u.Email)
		return www.UserError{
			ErrorCode: www.ErrorStatusRequiresTOTPCode,
		}
	}

	// Get the generated totp code. The provided code must match this
	// generated code.
	requestTime := time.Now()
	currentCode, err := p.totpGenerateCode(u.TOTPSecret, requestTime)
	if err != nil {
		return fmt.Errorf("totpGenerateCode: %v", err)
	}

	// Verify the user does not have too many failed attempts for this
	// epoch.
	if len(u.TOTPLastFailedCodeTime) >= totpFailedAttempts {
		// The user has too many failed attempts. We must first verify
		// that the failed attempts are from this epoch before this is
		// considered to be an error. If the generated code from the
		// failed timestamp matches the generated code from the current
		// timestamp then we know the failures occurred during this epoch.
		failedTS := u.TOTPLastFailedCodeTime[len(u.TOTPLastFailedCodeTime)-1]
		oldCode, err := p.totpGenerateCode(u.TOTPSecret, time.Unix(failedTS, 0))
		if err != nil {
			return fmt.Errorf("totpGenerateCode: %v", err)
		}
		if oldCode == currentCode {
			// The failures occurred in the same epoch which means the user
			// has exceeded their max allowed attempts.
			return www.UserError{
				ErrorCode: www.ErrorStatusTOTPWaitForNewCode,
			}
		}

		// Previous failures are not from this epoch. Clear them out.
		u.TOTPLastFailedCodeTime = []int64{}
	}

	// Verify the provided code matches the generated code
	var replyError error
	if currentCode == code {
		// The code matches. Clear out all previous failed attempts
		// before returning.
		u.TOTPLastFailedCodeTime = []int64{}
	} else {
		// The code doesn't match. Save the failure and return an error.
		ts := requestTime.Unix()
		u.TOTPLastFailedCodeTime = append(u.TOTPLastFailedCodeTime, ts)
		replyError = www.UserError{
			ErrorCode: www.ErrorStatusTOTPFailedValidation,
		}
	}

	// Update the user database
	err = p.db.UserUpdate(*u)
	if err != nil {
		return fmt.Errorf("UserUpdate: %v", err)
	}

	return replyError
}
```

Reply:
Yah that's much cleaner to read.